app/controllers/observations/images_controller.rb

```diff
@@ -0,0 +1,369 @@
+          process_image(args, params[:upload]["image#{i}"])
+          i += 1
+        end
+      end
+      redirect_with_query(permanent_observation_path(id: @observation.id))
+    end
+
+    def process_image(args, upload) # rubocop:disable Metrics/AbcSize
+      return if upload.blank?
+
+      @image = Image.new(args.permit(whitelisted_image_args))
+      @image.created_at = Time.current
+      @image.updated_at = @image.created_at
+      @image.user       = @user
+      @image.image      = upload
+      @image.save
+
+      # The 1st save (or !save) puts the image's original filename in the db,
+      # whether or not the user wants it.  So if we don't want it,
+      # we must empty it and save a 2nd time.
+      @image.original_name = "" if @user.keep_filenames == "toss"
+      return flash_object_errors(@image) unless @image.save
+
+      return revert_image_name_and_flash_errors unless
+        @image.process_image(strip: @observation.gps_hidden)
+
+      add_image_to_observation!
+    end
+
+    def add_image_to_observation!
+      @observation.add_image(@image)
+      @image.log_create_for(@observation)
+      name = @image.original_name
+      name = "##{@image.id}" if name.empty?
+      flash_notice(:runtime_image_uploaded_image.t(name: name))
+      update_related_projects(@image, params[:project])
+    end
+
+    def revert_image_name_and_flash_errors
+      name = @image.original_name
+      name = "???" if name.empty?
+      flash_error(:runtime_image_invalid_image.t(name: name))
+      flash_object_errors(@image)
+    end
+
+    def update_related_projects(img, checks)
+      return false unless checks
+
+      # Here's the problem: User can add image to obs he doesn't own
+      # if it is attached to one of his projects.
+      # Observation can be attached to other projects, too,
+      # though, including ones the user isn't a member of.
+      # We want the image to be attached even to these projects by default,
+      # however we want to give the user the ability NOT to attach his images
+      # to these projects which he doesn't belong to.
+      # This means we need to consider checkboxes not only of  user's projects,
+      # but also all  projects of the observation, as well.  Once it is detached
+      # from one of these projects the user isn't on,
+      # the checkbox will no longer show
+      # up on the edit_image form, preventing a user from attaching images to
+      # projects she doesn't belong to...
+      # except in the very strict case of uploading images for
+      # an observation which belongs to a project he doesn't belong to.
+      projects = @user.projects_member
+      img.observations.each do |obs|
+        obs.projects.each do |project|
+          projects << project unless projects.include?(project)
+        end
+      end
+
+      attach_images_to_projects_and_flash_notices(img, projects, checks)
+    end
+
+    def attach_images_to_projects_and_flash_notices(img, projects, checks)
```

Comment:
Add comment about what this method returns
```suggestion
    # Returns true if any changes made, false if none
    def attach_images_to_projects_and_flash_notices(img, projects, checks)
```